functions/api/category.js

```diff
@@ -0,0 +1,386 @@
+        categoryInfo[constants.IS_ACTIVE] = categoryData.isActive,
+        categoryInfo[constants.CREATED_DATE] = categoryData.createdDate.toDate(),
+        categoryInfo[constants.LAST_UPDATED_DATE] = categoryData.lastUpdatedDate.toDate()
+        categories.categories.push(categoryInfo);
+    })
+    categories[constants.TOTAL_CATEGORIES] = snapshot.size;
+    logger.debug('Returning categories to client.');
+    response.status(200).send(categories);
+});
+
+/**
+ * @description Route to retrieve single category data from firestore
+ * @returns Json object containing requested category
+ * @throws 400 if the product does not exists in firestore
+ */
+router.get('/:category', async (request, response, next) => {
+    var  requestedCategory = request.params.category.toLocaleLowerCase()
+    logger.info(`Retrieving category ${requestedCategory} from firestore`)
+    var categoryInfo = {}
+    const doc = db.collection(constants.CATEGORY).doc(requestedCategory);
+    const category = await doc.get()
+    if (!category.exists) {
+        const error = new Error(`Requested category ${requestedCategory} is not present in Firestore.`)
+        error.statusCode = 404
+        next(error)
+        return;
+    }
+    var categoryData = category.data()
+    categoryInfo[constants.CATEGORY] = category.id
+    categoryInfo[constants.DESCRIPTION] = categoryData.description
+    categoryInfo[constants.IS_ACTIVE] = categoryData.isActive
+    categoryInfo[constants.CREATED_DATE] = categoryData.createdDate.toDate()
+    categoryInfo[constants.LAST_UPDATED_DATE] = categoryData.lastUpdatedDate.toDate()
+    logger.debug(`Returning details for category ${requestedCategory} to client.`);
+    response.status(200).send(categoryInfo);
+});
+
+/**
+ * @description Route to retrieve all products from a given category
+ * @returns Json object containing array of products for a given category
+ */
+router.get('/products/:category', async (request, response, next) => {
+    var  requestedCategory = request.params.category.toLocaleLowerCase()
+    logger.info(`Retrieving products for category ${requestedCategory}`)
+    var categoryProducts = []
+    const doc = db.collection(constants.CATEGORY).doc(requestedCategory);
+    const category = await doc.get()
+    if (!category.exists) {
+        const error = new Error(`Requested category ${requestedCategory} is not present in Firestore.`)
+        error.statusCode = 404
+        next(error)
+        return;
+    }
+    var categoryData = category.data()    
+    categoryProducts = categoryData.products;
+    logger.debug(`Returning products from category ${requestedCategory} to client.`);
+    response.status(200).send(categoryProducts);    
+});
+
+/**
+ * @description Route to retireve all active/inactive categories from firestore
+ * @returns Json object containing all categories
+ */
+router.get("/:active/categories", async (request, response) => {
```

Comment:
if am not wrong, the route would be /api/categories/active/categories. It does'nt feel right.

Reply:
Yes, the first /api/categories is in route.js. All the urls starting with /api/categories will be routed to this method. Same for other /api/products/ will be routed to product.js.

Reply:
Umm, i meant the url is not looking right for this route. /api/**categories**/active/**categories** - 2 categories :)